feat(dev server): Initial dev server implementation

csojinb opened the pull request:
I have moved my dev server implementation over from https://github.com/falconry/falcon/pull/517. The initial status is basically the exact same implementation I put up for review over there. This is not intended to be mergeable yet, so feedback on design is most welcome. :smile:

The current functionality includes:
- auto-reloading
- legible tracebacks on exception (replacing werkzeug's fancy in-browser debugger, since that's less relevant for an API)
- option to drop into pdb debugger on exception

~~The current usage is Flask-like:~~
Modified to support command-line invocation:

```
# mything.py
import falcon

api = falcon.API()
```

```
$ updraft mything api
```

But, I like @kgriffs's (suggestion)[https://github.com/falconry/falcon/pull/517#issuecomment-114897903] to make it a command-line util, rather than having to make a py-file (`start.py` in the example above) to run at the command-line.

Ordinarily, I would not advocate for stuffing a first implementation with a whole bunch of features, but since we're starting from werkzeug, we could probably pick up some others without too much fuss. Other features I'm considering pulling in at the moment: 
- `extra_files`: files to watch outside of the application, e.g. config files, for triggering reloading
- threading: handling each request in a separate thread
- multiple processes

TODO:
- [x] add tests
- [x] handle licensing stuff for cannibalized werkzeug code
- [x] figure out which features of werkzeug's simple server we want to keep, other than the reloader, if any
- [x] strip down `serving.py` to only include what we need
- [x] ~~use `ipdb` if available~~ this might be difficult... the interfaces don't seem to be exactly the same
- [x] modify usage so that the server can be run from the command-line
- [ ] add `extra_files` support
- [ ] add basic usage docs

csojinb committed:
Add super-minimal setup.py

csojinb committed:
Add initial implementation of updraft dev server with reloading

- Adapted from werkzeug's "simple server"
- Only "simple" reloader is used. Werkzeug's Watchdog-based
reloader is not implemented
- Apache license for the repo as a whole, with BSD copyright/licence
notice and references for files that are adapted from Werkzeug. Not
sure if this part is done correctly.
- Known flakiness in test that verifies reloader behavior.
Papered over for the time being with flaky, but I'm not sure
why it's happening. Since the reloader comes from Werkzeug
(albeit in modified form), which is pretty stable, it seems
like it might be a problem with the test setup, but I never
pinned down the cause.

csojinb committed:
Run updraft as a command-line tool

- Just uses positional arguments for module and api names, rather
than gunicorn-like colon notation
- With tests for arg-parsing!
- Allows passing options to toggle reloader and debugger. Does not
allow configuration of reload interval or debugger method from
command line.

csojinb commented on updraft/__main__.py:
```diff
@@ -0,0 +1,32 @@
+    module_name = args.pop('module')
+    api_name = args.pop('api')
+
+    exec('from {} import {} as api'.format(module_name, api_name))
```
I don't like this, but I don't know if there's a better way to do this?

csojinb commented:
@kgriffs would you object to using pytest instead of nose for testing here? I'm working on setting up some test fixtures to test the server, and finding myself implementing stuff to spawn a subprocess in which to run the server from scratch. While interesting, it seems a bit fraught.

I haven't found anything for nose that makes spinning off subprocesses easy, but there is pytest-xprocess. (Also, I like the pytest fixtures a lot better than the nose ones. :smile:)

Open to other ideas though.

csojinb commented:
Also, I'm looking at the tagline for updraft again and realizing that my assumption that a falcon-dependence is expected might be wrong. Are we aiming to make this a framework-agnostic WSGI development server? If so, I'm going to need to rethink a lot of the implementation details here.

yohanboniface commented:
@csojinb I'd suggest you to release this as a Falcon plugin instead of waiting for an official review and endorsement, so people can start using it, giving feedback, patching it with PRs, etc. And when it get mature it can turn to something official :)

csojinb commented:
Okay, I'm looking at this again after a looong time, and I wanted to leave some notes on the current state:

1) I tried to add a test of the reloader itself, and it's flaky (see https://github.com/falconry/updraft/pull/1/commits/9e6aecbd858eb66fb8119aa0562cac10169f5abe). Werkzeug didn't actually have a test for this situation, perhaps for this reason. Never figured out why it flakes, but maybe it has something to do with subprocesses in subprocesses?
2) I can run the tests with py.test in my local virtualenv just fine (though I haven't updated the dependencies since I last worked on this many months ago), but I haven't gotten tox figured out yet. Some of the tests won't pass with tox (for reasons unknown to me), and there are some invocation errors that pop up that I never figured out.
3) I did a fair amount of work to simplify the Werkzeug code and get rid of pieces of it that we aren't using. (Perhaps it will get built back up like that too, but it seemed good to start with something simple.) I do not remember how much more of that stuff is still there.

csojinb committed:
Return nice tracebacks in response on 500

- Catches otherwise uncaught errors and returns 500
- Gives a traceback in the response
- Test verifies above behaviors
- Trackback also appears in server log
- This functionality occurs by default and has no configurable "off"
switch, which means that this server should NEVER BE USED IN PROD

csojinb committed:
Add option to drop into debugger on error

- Defaults to pdb.post_mortem, but another debug function can be
dropped in (e.g. ipdb, other debugger)
- Has test

csojinb commented:
@jmvrbanac @kgriffs I went through and cleaned up the git history a bit, and I think it should read fairly simply now.

This PR now has:
- Simple server with auto-reload option
- Application errors are caught and returned as 500s, with nicely-formatted tracebacks. (This is not configurable, in the current state. Another reason to prevent people from using this in prod.)
- Option to drop into debugger on application error. Defaults to pdb.post_mortem, but a different function can be passed into the debugger middleware via the run_dev_server entrypoint
- Apache license, with BSD copyright notice and license copied from Werkzeug and file-level comments on each file that was adapted from Werkzeug

This PR is missing:
- Full support for extra-file watching. There's capability in the reloader to specify extra files to watch to trigger reloading (e.g. for configuration files), as it default just looks at .py files, but that is not plumbed through to the updraft interface.
- Command-line support for configuring reloader interval (probably unnecessary) or custom debug method (dunno how to do that, maybe unnecessary)
- Completely deterministic tests. As mentioned above, the test of the reloader flakes and has been annotated with @flaky to (slightly) improve the success rate. I still think this is likely a problem with the test setup, but I could be wrong.
- Documentation of any kind.
- Continuous integration
- Setup for testing against multiple Python versions. (tests run locally with Python 2.7.10)

csojinb commented:
Okay, I also separated out the original work from the werkzeug-adapted stuff more clearly for the serving tests.
